### PR TITLE
NIFI-5252 - support arbitrary headers in PutEmail processor

### PR DESCRIPTION
**Pull request by dtrodrigues**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutEmail.java [hunk 1]
```diff
@@ -319,6 +330,15 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             message.setRecipients(RecipientType.CC, toInetAddresses(context, flowFile, CC));
             message.setRecipients(RecipientType.BCC, toInetAddresses(context, flowFile, BCC));
 
+            final String attributeNameRegex = context.getProperty(ATTRIBUTE_NAME_REGEX).getValue();
+            final Pattern attributeNamePattern = attributeNameRegex == null ? null : Pattern.compile(attributeNameRegex);
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutEmail.java]: Could we move the compile outside the onTrigger call? I think it's better to make this call when the processor is scheduled.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutEmail.java [hunk 2]
```diff
@@ -319,6 +330,15 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+            if (attributeNamePattern != null) {
+                for (final Map.Entry<String, String> entry : flowFile.getAttributes().entrySet()) {
+                    if (attributeNamePattern.matcher(entry.getKey()).matches()) {
```
Comment (ottobackwards) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutEmail.java]: There are rules about how the headers have to be encoded.  We should use the MimeUtility to ensure everything is encoded correctly.

```
Note that RFC 822 headers must contain only US-ASCII characters, so a header that contains non US-ASCII characters must have been encoded by the caller as per the rules of RFC 2047.
```

[MimeUtility](https://docs.oracle.com/javaee/6/api/javax/mail/internet/MimeUtility.html)

We should also have tests for ensuring attributes with content that must be encoded are handled.

**Comment by dtrodrigues** [comment]
moved regex compilation to when processor is scheduled and ensured header values are encoded appropriately

**Comment by pvillard31** [comment]
LGTM - @ottobackwards can you have a look as well? if all looks good for you, will merge to master. Thanks guys!

**Comment by ottobackwards** [comment]
The changes to the encoding looks good.  
Thanks for the contribution!

+1

**Pull request closed without merging**